### ISSUE TITLE
Local index addition for distributed sparse domains

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#12673]
`+=` and `bulkAdd` interface can be used to add indices to block distributed
sparse domains.

```chapel
var indices = [{1..numIndices}] idx_type;
populate(indices);
distSparseDom += indices;
```

Today, this ends up `indices.locale` sorting the array using a special
comparator so that the indices are sorted for their target locale, and each
locale doing their own inside a `coforall`.  The main problem I see with this
status quo is that it is centralized, therefore arguably not scalable.

This issue is to propose a local index addition for sparse domains. There are
some cases where locales can compute/generate/read their own indices. One
specific example is the Sparse benchmark in PRK. Another more abstract example
is simply storing very large sparse data in a file system.

What I personally have in mind would look something like this:

```chapel
coforall l in Locales do on l {
  var indices = [{1..numIndices}] idx_type;
  populateMyLocalIndices(indices, here);
  sparseDom.bulkAddHere(indices);
}
```

One question that popped in my mind was what happens if an index in indices
doesn't really belong to the particular locale calling `bulkAddHere`. I think it
is analogous to adding to a non-distributed sparse domain an index that is not a
member of the parent domain. Which should result in error under `boundsCheck`
and undefined behavior without it.

**Comment by e-kayrakli** [comment]
In terms of the method name; another alternative I can think of is `localBulkAdd`. There are instances of both `*Here` methods and `local*` methods and functions in modules currently. I have affinity (pun intended) to `bulkAddHere` as it reads more like a sentence. I wouldn't lose sleep if `localBulkAdd` has more support.

**Comment by e-kayrakli** [comment]
This is something small that I can take up on. Is there any objection @bradcray @mppf?

**Comment by bradcray** [comment]
I'm open to this.

**Comment by benharsh** [comment]
Would it make sense to have this method (whatever we call it) accept an optional ``locale`` argument indicating where the addition should take place?

Going a bit further (perhaps too far): What if we had a special locale value indicating "any locale", or maybe "unknown locale"? We have something sort of like this in NUMA with child locales called ``c_sublocid_any``. In the context of our bulk-add method(s) could this special value indicate that the method should figure out the correct locales itself, and otherwise will add indices to the specified locale? Does that appeal to anyone?

**Comment by e-kayrakli** [comment]
@benharsh I am making sure that I understand your suggestions:

> Would it make sense to have this method (whatever we call it) accept an optional locale argument indicating where the addition should take place?

I like this. Just to be clear it is going to fire an `on` on the desired locale, copy the indices in bulk, and do the index addition there. Currently with what I am proposing, it would be done as follows:

```chapel
var indicesForLocale1 = generateIndicesForLocale1();
on Locales[1] {
  var localIndices: [indicesForLocale1.domain] int = indicesForLocale1;
  sparseDom.bulkAddHere(localIndices);
}
```

With what you are suggesting the application code will look more like:

```chapel
var indicesForLocale1 = generateIndicesForLocale1();
sparseDom.localBulkAdd(localIndices, Locales[1]);
```

If that's the case, I am leaning more towards the method name `localBulkAdd` because it makes more sense for a method like this to accept a locale argument.

> Going a bit further (perhaps too far): What if we had a special locale value indicating "any locale", or maybe "unknown locale"? We have something sort of like this in NUMA with child locales called c_sublocid_any. In the context of our bulk-add method(s) could this special value indicate that the method should figure out the correct locales itself, and otherwise will add indices to the specified locale? Does that appeal to anyone?

This one I am not entirely following. Currently there is already a `bulkAdd` for distributed sparse that figures out where each index will go. Then, what you are suggesting becomes adding an optional locale argument to `bulkAdd` which basically controls everything we are discussing here. Is that right? I think that's also a good idea. This way we will have added only an argument to a method, instead of adding a completely new method to the interface.

So there would be:

```chapel
//figure out where each index will go and do the addition there (we have this already)
bulkAdd(indices);   
// hey, all the indices here belong to Locales[1], so go do the addition only there
bulkAdd(indices, loc=Locales[1]);   
```

Then, maybe `bulkAddHere` can be added only as a convenience wrapper for `bulkAdd(indices, loc=here);`

**Comment by benharsh** [comment]
> Then, what you are suggesting becomes adding an optional locale argument to bulkAdd which basically controls everything we are discussing here. Is that right?

Yes, that's correct. This also depends on a useful default value for the argument, and I'm not sure that one exists because as far as I know we currently do not have locale values other than Locale0..#LocaleN.

**Comment by e-kayrakli** [comment]
If there is no workaround for that, then we can overload `bulkAdd`.

(Can't a default argument be `nil`? Probably this is something related to @mppf's nillable types work)

**Comment by e-kayrakli** [comment]
Closed via #13411
